File: job-2022-autumn/42.trapping-rain-water.js
/*
 * @lc app=leetcode id=42 lang=javascript
 *
 * [42] Trapping Rain Water
 */

// @lc code=start
/**
 * @param {number[]} height
 * @return {number}
 */
var trap = function (height) {
    const n = height.length;
    const maxLeft = new Array(n);
    const maxRight = new Array(n);

    maxLeft[0] = height[0];
    maxRight[height.length - 1] = height[height.length - 1];

    for (let i = 1; i < height.length; i++) {
        maxLeft[i] = Math.max(height[i], maxLeft[i - 1]);
        maxRight[n - 1 - i] = Math.max(
            height[n - 1 - i],
            maxRight[n - 1 - i + 1],
        );
    }

    let water = 0;

    for (let i = 0; i < n; i++) {
        water += Math.min(maxLeft[i], maxRight[i]) - height[i];
    }

    return water;
};
// @lc code=end

File: Daily/42.trapping-rain-water.cpp
/*
 * @lc app=leetcode id=42 lang=cpp
 *
 * [42] Trapping Rain Water
 */
#include <cmath>
#include <iostream>
#include <vector>

using std::vector;

// @lc code=start
class Solution {
 public:
  int trap(const vector<int>& heights) {
    const int N = heights.size();
    vector<int> maxLeftHeight(N);
    vector<int> maxRightHeight(N);

    maxLeftHeight[0] = heights[0];
    maxRightHeight[N - 1] = heights[N - 1];

    for (int i = 1; i < N; i++) {
      maxLeftHeight[i] = std::max(maxLeftHeight[i - 1], heights[i]);
      maxRightHeight[N - i - 1] =
          std::max(maxRightHeight[N - i], heights[N - i - 1]);
    }

    int water = 0;
    for (int i = 0; i < N; i++) {
      water += std::min(maxLeftHeight[i], maxRightHeight[i]) - heights[i];
    }

    return water;
  }
};
// @lc code=end